src/PotentialSimulation/ConvergenceAndRefinement.jl

```diff
@@ -88,282 +88,181 @@ function _update_till_convergence!( fssrb::PotentialSimulationSetupRB{T, N1, N2}
+                ns[ia][i] -= 1
+                sub_widths[ia][i] = widths[ia][i] / (ns[ia][i]+1)
             end
         end
     end
-
-    if isodd(length(inds_x))
-        for ix in 1:size(potential, 1)
-            if !in(ix, inds_x)
-                if ix != size(potential, 1)
-                    if ax_x[ix + 1] - ax_x[ix] > minimum_distance_x
-                        append!(inds_x, ix)
-                        break
-                    end
-                end
-            end
+    for i in 1:3 # always add an even number of ticks
+        if isodd(sum(ns[i])) 
+            i_max_width = findmax(sub_widths[i])[2]
+            ns[i][i_max_width] += 1 
+            sub_widths[i][i_max_width] = widths[i][i_max_width] / (ns[i][i_max_width]+1)
         end
     end
-    if isodd(length(inds_x))
-        b = true
-        diffs = diff(ax_x)
-        while b
-            if length(diffs) > 0
-                idx = findmax(diffs)[2]
-                if !in(idx, inds_x)
-                    append!(inds_x, idx)
-                    b = false
-                else
-                    deleteat!(diffs, idx)
-                end
-            else
-                for i in eachindex(ax_x[1:end-1])
-                    if !in(i, inds_x) push!(inds_x, i) end
-                end
-                b = false
-            end
+    new_axes = broadcast(i -> _refine_axis(p.grid.axes[i], ns[i], sub_widths[i]), (1, 2, 3))
+    return typeof(p.grid)(new_axes)
+end
+
+function _refine_axis(ax::DiscreteAxis{T, <:Any, <:Any, ClosedInterval{T}}, ns::Vector{Int}, sub_widths::Vector{T}) where {T, I}
+    @assert length(ns) == length(ax.ticks)-1 # for ClosedInterval axis
+    ticks = Vector{T}(undef, length(ax.ticks) + sum(ns))
+    i = 1 
+    for j in eachindex(ns)
+        ticks[i] = ax.ticks[j]
+        for k in 1:ns[j]
+            i += 1 
+            ticks[i] = ax.ticks[j] + k * sub_widths[j]
         end
+        i += 1 
     end
-    if isodd(length(inds_x)) inds_x = inds_x[1:end-1] end
-    @assert iseven(length(inds_x)) "Refinement would result in uneven grid in x. This is not allowed since this is the red black dimension."
-    
-
-    return sort!(inds_x), sort!(inds_y), sort!(inds_z)
+    ticks[end] = ax.ticks[end]
+    typeof(ax)(ax.interval, ticks)
 end
 
-function refine(p::ScalarPotential, max_diffs::Tuple{<:Real,<:Real,<:Real}, minimum_distances::Tuple{<:Real,<:Real,<:Real})::typeof(p) 
-    T = eltype(p.grid)
-    refine_at_inds = _get_refinement_inds(p.data, p.grid, T.(max_diffs), T.(minimum_distances))
-    potential, grid = add_points_and_interpolate(p.data, p.grid, refine_at_inds...)
-    return typeof(p)(potential, grid)
-end
\ No newline at end of file
+
+function _extend_refinement_limits(rl::Union{<:Real, Vector{<:Real}, Tuple{<:Real,<:Real,<:Real}, Vector{<:Tuple{<:Real, <:Real, <:Real}}})
+    0
+end
+_extend_refinement_limits(rl::Real) = (rl, rl, rl )
+_extend_refinement_limits(rl::Tuple{<:Real,<:Real,<:Real}) = rl

```

Comment:
`_extend_refinement_limits` is defined twice for `<:Real` (L.265/267) and `Tuple[<:Real, <:Real, <:Real}` (L.265/268). Is this intended?

Reply:
Nope. I removed the wrong one.